src/lib.rs

```diff
@@ -0,0 +1,736 @@
+            uattributes.permission_level = Some(permission_level);
+        }
+        if let Some(commstatus) = attachment.get(&"commstatus".as_bytes()) {
+            let commstatus = String::from_utf8(commstatus.to_vec())?.parse::<i32>()?;
+            uattributes.commstatus = Some(commstatus);
+        }
+        if let Some(reqid) = attachment.get(&"reqid".as_bytes()) {
+            let reqid = UUID::parse_from_bytes(&reqid)?;
+            uattributes.reqid = Some(reqid).into();
+        }
+        if let Some(token) = attachment.get(&"token".as_bytes()) {
+            let token = token.to_string();
+            uattributes.token = Some(token);
+        }
+        if let Some(traceparent) = attachment.get(&"traceparent".as_bytes()) {
+            let traceparent = traceparent.to_string();
+            uattributes.traceparent = Some(traceparent);
+        }
+        Ok(uattributes)
+    }
+
+    async fn send_publish(
+        &self,
+        zenoh_key: &str,
+        payload: UPayload,
+        attributes: UAttributes,
+    ) -> Result<(), UStatus> {
+        // Get the data from UPayload
+        let Some(Data::Value(buf)) = payload.data else {
+            // TODO: Assume we only have Value here, no reference for shared memory
+            return Err(UStatus::fail_with_code(
+                UCode::INVALID_ARGUMENT,
+                "Invalid data",
+            ));
+        };
+
+        // Serialized UAttributes into protobuf
+        let Ok(attachment) = UPClientZenoh::uattributes_to_attachment(&attributes) else {
+            return Err(UStatus::fail_with_code(
+                UCode::INVALID_ARGUMENT,
+                "Invalid uAttributes",
+            ));
+        };
+
+        let priority =
+            UPClientZenoh::map_zenoh_priority(attributes.priority.enum_value().map_err(|_| {
+                UStatus::fail_with_code(UCode::INVALID_ARGUMENT, "Invalid priority")
+            })?);
+
+        let putbuilder = self
+            .session
+            .put(zenoh_key, buf)
+            .encoding(Encoding::WithSuffix(
+                KnownEncoding::AppCustom,
+                payload.format.value().to_string().into(),
+            ))
+            .priority(priority)
+            .with_attachment(attachment.build());
+
+        // Send data
+        putbuilder
+            .res()
+            .await
+            .map_err(|_| UStatus::fail_with_code(UCode::INTERNAL, "Unable to send with Zenoh"))?;
+
+        Ok(())
+    }
+
+    async fn send_response(
+        &self,
+        zenoh_key: &str,
+        payload: UPayload,
+        attributes: UAttributes,
+    ) -> Result<(), UStatus> {
+        // Get the data from UPayload
+        let Some(Data::Value(buf)) = payload.data else {
+            // TODO: Assume we only have Value here, no reference for shared memory
+            return Err(UStatus::fail_with_code(
+                UCode::INVALID_ARGUMENT,
+                "Invalid data",
+            ));
+        };
```

Comment:
FMPOV it is not mandatory for a message to have a payload, is it?
What about an RPC *turn_on_hvac* which simply returns with a successful comm_status?